### PR TITLE
math.c: hypot

### PR DESCRIPTION
**Pull request by nobu**
- math.c (math_hypot): extend to arbitrary number of arguments.

### CODE REVIEW

File: math.c [hunk 1]
```diff
@@ -681,15 +681,35 @@ math_ldexp(VALUE obj, VALUE x, VALUE n)
...
 
 static VALUE
-math_hypot(VALUE obj, VALUE x, VALUE y)
+math_hypot(int argc, VALUE *argv, VALUE obj)
+{
+    int i;
+    double x, y;
+    rb_check_arity(argc, 1, UNLIMITED_ARGUMENTS);
+    x = Get_Double(argv[0]);
+    if (argc == 1) return DBL2NUM(x);
```
Comment (gogotanaka) [inline math.c]: Need to return the absolute value?

Comment (nobu) [inline math.c]: Exactly, thank you.

**Comment by akr** [comment]
It is better to avoid overflow for big arguments if the final result can be represented in double type.
For example, current implementation overflows if an argument is bigger than sqrt(DBL_MAX).

SUSv4 explains overflow of hypot() function as follows.
"These functions take precautions against overflow during intermediate steps of the computation."

ECMAScript 6, which has hypot() function with any number of arguments, explains as follows.
"Implementations should take care to avoid the loss of precision from overflows and underflows that are prone to occur in naive implementations when this function is called with two or more arguments."
http://www.ecma-international.org/ecma-262/6.0/

**Comment by nobu** [comment]
Interesting.

~~~ruby
x=(Float::MAX/Math.sqrt(2.0))
Math.hypot(x, x) #=> 1.7976931348623155e+308
Math.hypot(x, x, 0) #=> Infinity
~~~

How does `hypot(3m)` avoid overflow?

**Comment by akr** [comment]
It seems NetBSD's hypot() and its original, fdlibm's hypot() scales
too big numbers by `2**-600` and too small numbers by `2**600`.

http://cvsweb.netbsd.org/cgi-bin/cvsweb.cgi/src/lib/libm/src/e_hypot.c?rev=1.13&content-type=text/x-cvsweb-markup
http://www.netlib.org/fdlibm/e_hypot.c

**Pull request closed without merging**